### PR TITLE
arch/arm64: Implement ASID support in ARM64 MMU

### PR DESCRIPTION
**Pull request by hx-bora**
Improves a lot context-switch performance.

TLB Invalidating and the nG bit are used conservatively. This could be improved in future work.

Note: MMU configurations changes are known to be error prone. Please have someone review this patch very carefully.

### CODE REVIEW

**Comment by npitre** [comment]
Thanks for this work.

Some comments:

- Could you provide numbers in the commit log illustrating the context 
  switch improvements? For example, the output from tests/benchmarks/sched 
  before and after this change.

- Please define `next_asid` as a static variable.

- If you remove `is_ptable_active()` you should probably replace those
  `invalidate_tlb_all()` invocations that were conditional on 
  `is_ptable_active()` by an ASID-based TLB invalidate version, otherwise
  performance will likely become worse not better.

- This code:

```
        if ((next_asid >= (1U << VM_ASID_BITS)) || (next_asid == 0)) {
                next_asid = 1;
        }
```

  Testing against 0 is redundant as it will never happen.

File: arch/arm64/core/mmu.c [hunk 1]
```diff
@@ -1155,9 +1179,17 @@ int arch_mem_domain_thread_remove(struct k_thread *thread)
 static void z_arm64_swap_ptables(struct k_thread *incoming)
 {
 	struct arm_mmu_ptables *ptables = incoming->arch.ptables;
+	uint64_t curr_ttbr0 = read_sysreg(ttbr0_el1);
```
Comment (povergoing) [inline arch/arm64/core/mmu.c:1182]: please use `read_ttbr0_el1()`

**Comment by povergoing** [comment]
I am also curious about the improvements :)

**Comment by npitre** [comment]
Excellent.

If you could add your test code in tests/kernel/sched/ and then
include those results in the commit log for the ASID support 
that'd be perfect.

**Review by npitre — APPROVED** [review]

**Review by povergoing — APPROVED** [review]

**Review by carlocaione — APPROVED** [review]

**Review by SgrrZhf — APPROVED** [review]

**Pull request merged**